RFC-0001-native-spi.md

```diff
@@ -0,0 +1,548 @@
+
+Once this work is completed, we can create a `JavaWorkerSessionPropertiesManager` that manages purely Java worker session properties.  We can migrate existing Java worker-only session properties to this new plugin as and when they are identified.  This can be a default plugin to ensure continuity and prevent disruption in Presto users’ deployments.
+
+![Session property flow](RFC-0002/session.png)
+
+
+
+### Functions validation and execution
+
+As the source of truth will now be whatever functions are registered in Velox, the coordinator must have a master list of these functions that is consistent with whatever is registered, otherwise functions will parse correctly according to whatever is registered in Java, yet fail during execution.  Additionally, `SHOW FUNCTIONS` should simply show functions that are eligible for execution.
+
+
+#### How functions will be known to the coordinator
+
+Velox already registers all functions for execution.  Additionally, it’s already possible to obtain a list of functions from Velox.  In order to report these functions to the coordinator, a new endpoint will be added, `/v1/functions`.  A `GET` to this endpoint will return the total list of all functions registered in Velox in JSON format.  The particular format will be compatible with `UdfFunctionSignatureMap`, an existing JSON serialization format currently used to load function definitions from files.
+
+Presto can already define custom functions with the existing SPI interface `FunctionNamespaceManager`, which facilitates the execution of SQL functions and remote UDFs.  We will enhance this integration in the Presto engine to permit the execution of remote functions during constant folding.  A new implementation will be added, `NativeFunctionNamespaceManager`.  This manager will retrieve the list of functions from the sidecar’s `/v1/functions` endpoint.
+
+Functions returned from this endpoint should behave as built-in functions to the user.  Built-in functions are currently simply functions that don’t require a catalog and schema to be specified when they are executed.  Because each `FunctionNamespaceManager` registers itself to the coordinator with a catalog and namespace to refer, we must allow the functions from the `NativeFunctionNamespaceManager` to be addressed as if they were built-in functions.  To support this, the built-in namespace will be become configurable, and a `FunctionNamespaceManager` can indicate to the coordinator that its own namespace should be considered the default namespace.  The coordinator will validate that only one such plugin has specified itself as a default namespace.  The `NativeFunctionNamespaceManager` will indicate that it is a default namespace, and hence will replace built-in functions in Java.
+
+![Function flow](RFC-0002/functions.png)
+
+
+
+#### How will constant folding work
```

Comment:
@tdcmeehan Would it be possible to add the explanation what is exactly constant folding here and why it should be treated as normal functions? It'll be nice to give an example.

Reply:
Done, PTAL!